README.md
This package provides a library for representing local dates that are
independent of the time zone.

A date of birth represented as a special case of local date with ability
to determine the age.

## Features

* Representation of local date (see `LocalDate` API):
  * Creation from `DateTime`;
  * Conversion to `DateTime`;
  * Generic API for implementation of different formatting;
  * Generic API for implementation of different format parsing;
  * Support of formatting/parsing of ISO8601 representation of date;
  * Comparison of local dates.
* Representation of a date of birth as a special case of local date (see `DateOfBirth` API):
  * Calculation of the age on the given date with handling of leap date of birth;
  * Determination of a birthday date in the specified year with handling of leap date of birth.

## Getting started

1. Install the package as a dependency:

    ```shell
    dart pub add local_date
    ```

2. Make the import:

    ```dart
    import 'package:local_date/local_date.dart';
    ```

## Usage

### Parsing / formatting

Parse ISO 8601 string representation:

```dart
final date = LocalDate.parse('2023-11-15', parser: ISO8610Format());
```

Format as ISO 8601 string:

```dart
final date = LocalDate.of(2023, 11, 15);

assert(date.formattedBy(ISO8601Format()) == '2023-11-15');
```

Feel free to implement `DateFormatter` and `DateParser` generic interfaces
for custom representations.

### Working with dates of birth

Interpretation of the date of birth is not so straightforward as most of
us used to think about.

One of the most widespread mistake is a wrong interpretation of leap dates
of birth. If a person has date of birth on February 29th, the birthday in
a non-leap year must be February 28th, not March 1st.

Fortunately, this library encapsulates correct logic inside `DateOfBirth`.

Let's see the example:

```dart
final leapDateOfBirth = DateOfBirth.of(2000, 2, 29);

// Age will be incremented on since February 28 in a non-leap lear:
assert(leapDateOfBirth.calculateAgeAsOnDate(LocalDate.of(2003, 2, 28)) == 3);

// Age will be incremented on since February 29 in a leap lear:
assert(leapDateOfBirth.calculateAgeAsOnDate(LocalDate.of(2004, 2, 28)) == 3);
assert(leapDateOfBirth.calculateAgeAsOnDate(LocalDate.of(2004, 2, 29)) == 3);
assert(leapDateOfBirth.calculateAgeAsOnDate(LocalDate.of(2004, 12, 31)) == 3);
```

## Additional information

### Limitations

This implementation works with the Gregorian calendar only and limits set
of years to the range from 0 to 9999. This range covers the most of
the enterprise related issues.


CHANGELOG.md
## 0.1.0+1

- Fixes analyzer warnings for the package internals.

## 0.1.0

- Initial version.


test/year_test.dart
// Copyright (c) 2023 LLC "LitGroup"
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is furnished
// to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

import 'package:test/test.dart';

import 'package:local_date/local_date.dart';

void main() {
  group('Year', () {
    // Construction
    //--------------------------------------------------------------------------
    group('construction from int', () {
      test('succeeds', () {
        expect(() => Year.from(0), returnsNormally,
            reason: 'Min supported value is 0.');
        expect(() => Year.from(9999), returnsNormally,
            reason: 'Max supported value is 9999');
      });

      test('fails if value is out of the supported range', () {
        expect(() => Year.from(0 - 1), throwsArgumentError,
            reason: 'Min supported value is 0.');
        expect(() => Year.from(9999 + 1), throwsArgumentError,
            reason: 'Max supported value is 9999.');
      });
    });

    group('failable creation from int', () {
      test('succeeds', () {
        expect(Year.tryFrom(0), equals(Year.from(0)));
        expect(Year.tryFrom(9999), equals(Year.from(9999)));
      });

      test('results to null if value is out of the supported range', () {
        expect(Year.tryFrom(0 - 1), isNull);
        expect(Year.tryFrom(9999 + 1), isNull);
      });
    });

    //--------------------------------------------------------------------------

    group('leap year determining:', () {
      const dataSet = [
        (1999, isLeap: false, reason: 'not divisible by 4'),
        (2024, isLeap: true, reason: 'divisible by 4, but not by 100'),
        (1700, isLeap: false, reason: 'divisible by 4 and 100, but not by 400'),
        (1600, isLeap: true, reason: 'divisible by 4, 100 and 400'),
      ];

      for (final (yearNumber, :isLeap, :reason) in dataSet) {
        test('$yearNumber is ${isLeap ? 'leap' : 'not leap'} ($reason)', () {
          final year = Year.from(yearNumber);

          expect(year.isLeap, isLeap);
          expect(year.isNotLeap, !isLeap);
        });
      }
    });

    test('determining of a duration (in days)', () {
      final leapYear = Year.from(2000);
      expect(leapYear.durationInDays, equals(366),
          reason: 'Duration of a leap year is 366 days.');

      final nonLeapYear = Year.from(2001);
      expect(nonLeapYear.durationInDays, equals(365),
          reason: 'Duration of a non-leap year is 365 days.');
    });

    // Comparison
    //--------------------------------------------------------------------------

    test('equality & hashing', () {
      final year = Year.from(2000);
      final sameYear = Year.from(2000);

      final yearBefore = Year.from(2000 - 1);
      final yearAfter = Year.from(2000 + 1);

      expect(year, equals(sameYear));
      expect(year.hashCode, equals(sameYear.hashCode),
          reason: 'Hash codes of equal years must be equal.');

      expect(year, isNot(equals(yearBefore)));
      expect(year, isNot(equals(yearAfter)));
    });

    test('comparison to another year', () {
      final year2000 = Year.from(2000);

      expect(year2000.compareTo(Year.from(2000)), isZero);
      expect(year2000.compareTo(Year.from(2000 - 1)), isPositive);
      expect(year2000.compareTo(Year.from(2000 + 1)), isNegative);
    });

    test('is after another year', () {
      final year2000 = Year.from(2000);

      expect(year2000.isAfter(Year.from(1999)), isTrue);
      expect(year2000.isAfter(Year.from(2000)), isFalse);
      expect(year2000.isAfter(Year.from(2001)), isFalse);
    });

    test('is before another year', () {
      final year2000 = Year.from(2000);

      expect(year2000.isBefore(Year.from(2001)), isTrue);
      expect(year2000.isBefore(Year.from(2000)), isFalse);
      expect(year2000.isBefore(Year.from(1999)), isFalse);
    });

    test('is the same as another year', () {
      final year2000 = Year.from(2000);

      expect(year2000.isSameAs(Year.from(2000)), isTrue);
      expect(year2000.isSameAs(Year.from(1999)), isFalse);
      expect(year2000.isSameAs(Year.from(2001)), isFalse);
    });

    // Conversion to a different type
    //--------------------------------------------------------------------------

    test('conversion to string', () {
      var year = Year.from(0);
      expect(year.toString(), equals('0000'));

      year = Year.from(23);
      expect(year.toString(), equals('0023'));

      year = Year.from(2023);
      expect(year.toString(), equals('2023'));
    });

    test('conversion to int', () {
      final year = Year.from(2023);
      expect(year.toInt(), equals(2023));
    });
  });
}


test/month_test.dart
// Copyright (c) 2023 LLC "LitGroup"
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is furnished
// to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

import 'package:test/test.dart';

import 'package:local_date/local_date.dart';

void main() {
  // Construction
  //----------------------------------------------------------------------------

  group('Month', () {
    group('construction from number', () {
      test('succeeds', () {
        expect(Month.from(1), Month.january);
        expect(Month.from(2), Month.february);
        expect(Month.from(3), Month.march);
        expect(Month.from(4), Month.april);
        expect(Month.from(5), Month.may);
        expect(Month.from(6), Month.june);
        expect(Month.from(7), Month.july);
        expect(Month.from(8), Month.august);
        expect(Month.from(9), Month.september);
        expect(Month.from(10), Month.october);
        expect(Month.from(11), Month.november);
        expect(Month.from(12), Month.december);
      });

      test('throws error for invalid number', () {
        expect(() => Month.from(0), throwsArgumentError);
        expect(() => Month.from(13), throwsArgumentError);
      });
    });

    group('failable construction from number', () {
      test('succeeds', () {
        expect(Month.tryFrom(1), Month.january);
        expect(Month.tryFrom(2), Month.february);
        expect(Month.tryFrom(3), Month.march);
        expect(Month.tryFrom(4), Month.april);
        expect(Month.tryFrom(5), Month.may);
        expect(Month.tryFrom(6), Month.june);
        expect(Month.tryFrom(7), Month.july);
        expect(Month.tryFrom(8), Month.august);
        expect(Month.tryFrom(9), Month.september);
        expect(Month.tryFrom(10), Month.october);
        expect(Month.tryFrom(11), Month.november);
        expect(Month.tryFrom(12), Month.december);
      });

      test('results to null for invalid number', () {
        expect(Month.tryFrom(0), isNull);
        expect(Month.tryFrom(13), isNull);
      });
    });

    // Enumerated values
    //--------------------------------------------------------------------------

    test('list of values', () {
      expect(Month.values, [
        Month.january,
        Month.february,
        Month.march,
        Month.april,
        Month.may,
        Month.june,
        Month.july,
        Month.august,
        Month.september,
        Month.october,
        Month.november,
        Month.december,
      ]);
    });

    // Number of days determining
    //--------------------------------------------------------------------------

    group('number of days determining', () {
      const testExamples = <(Month, int daysInNonLeapYear, int daysInLeapYear)>[
        (Month.january, 31, 31),
        (Month.february, 28, 29),
        (Month.march, 31, 31),
        (Month.april, 30, 30),
        (Month.may, 31, 31),
        (Month.june, 30, 30),
        (Month.july, 31, 31),
        (Month.august, 31, 31),
        (Month.september, 30, 30),
        (Month.october, 31, 31),
        (Month.november, 30, 30),
        (Month.december, 31, 31),
      ];

      final nonLeapYear = Year.from(2001);
      final leapYear = Year.from(2000);

      for (final (month, nonLeapLength, leapLength) in testExamples) {
        test('in $month using leap/non-leap year sign', () {
          expect(month.days(leapYear: false), equals(nonLeapLength));
          expect(month.days(leapYear: true), equals(leapLength));
        });

        test('in $month for the specified year', () {
          expect(month.daysInYear(nonLeapYear), equals(nonLeapLength));
          expect(month.daysInYear(leapYear), equals(leapLength));
        });
      }
    });

    // Comparison
    //--------------------------------------------------------------------------

    test('equality & hash code', () {
      expect(Month.august, equals(Month.august));
      expect(Month.august, equals(Month.from(8)));
      expect(Month.august, equals(Month.tryFrom(8)!));

      expect(Month.august.hashCode, equals(Month.from(8).hashCode),
          reason: 'Hash code of equal values must be equal too.');

      expect(Month.august, isNot(Month.july));
      expect(Month.august, isNot(Month.september));
    });

    test('comparison to another month', () {
      expect(Month.august.compareTo(Month.august), isZero);
      expect(Month.august.compareTo(Month.september), isNegative);
      expect(Month.august.compareTo(Month.july), isPositive);
    });

    test('is after another in calendar order', () {
      final april = Month.april;

      expect(april.isAfter(Month.march), isTrue);
      expect(april.isAfter(Month.april), isFalse);
      expect(april.isAfter(Month.may), isFalse);
    });

    test('is before another in calendar order', () {
      final april = Month.april;

      expect(april.isBefore(Month.may), isTrue);
      expect(april.isBefore(Month.april), isFalse);
      expect(april.isBefore(Month.march), isFalse);
    });

    test('is the same as another month', () {
      final april = Month.april;

      expect(april.isSameAs(Month.april), isTrue);
      expect(april.isSameAs(Month.march), isFalse);
      expect(april.isSameAs(Month.may), isFalse);
    });

    // Conversion to a different type
    //--------------------------------------------------------------------------

    test('conversion to string', () {
      expect(Month.january.toString(), 'January');
      expect(Month.february.toString(), 'February');
      expect(Month.march.toString(), 'March');
      expect(Month.april.toString(), 'April');
      expect(Month.may.toString(), 'May');
      expect(Month.june.toString(), 'June');
      expect(Month.july.toString(), 'July');
      expect(Month.august.toString(), 'August');
      expect(Month.september.toString(), 'September');
      expect(Month.october.toString(), 'October');
      expect(Month.november.toString(), 'November');
      expect(Month.december.toString(), 'December');
    });

    test('conversion to int', () {
      expect(Month.january.toInt(), 1);
      expect(Month.february.toInt(), 2);
      expect(Month.march.toInt(), 3);
      expect(Month.april.toInt(), 4);
      expect(Month.may.toInt(), 5);
      expect(Month.june.toInt(), 6);
      expect(Month.july.toInt(), 7);
      expect(Month.august.toInt(), 8);
      expect(Month.september.toInt(), 9);
      expect(Month.october.toInt(), 10);
      expect(Month.november.toInt(), 11);
      expect(Month.december.toInt(), 12);
    });
  });
}


test/local_date_test.dart
// Copyright (c) 2023 LLC "LitGroup"
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is furnished
// to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

import 'package:test/test.dart';

import 'package:local_date/local_date.dart';

import 'fixtures/date_test_data_set.dart';
import 'fixtures/test_date_format.dart';

void main() {
  group('LocalDate', () {
    // Construction tests
    //--------------------------------------------------------------------------

    group('failable creation from int values', () {
      for (final (index, example) in validDateExamples.indexed) {
        final ((year, month, day), reason) = example;
        test('succeeds #$index', () {
          expect(LocalDate.tryOf(year, month, day), isNotNull, reason: reason);
        });
      }

      for (final (index, example) in invalidDateExamples.indexed) {
        final ((year, month, day), reason) = example;
        test('returns null for invalida date #$index', () {
          expect(LocalDate.tryOf(year, month, day), isNull, reason: reason);
        });
      }
    });

    group('creation from int values', () {
      for (final (index, example) in validDateExamples.indexed) {
        final ((year, month, day), reason) = example;
        test('succeeds #$index', () {
          expect(() => LocalDate.of(year, month, day), returnsNormally,
              reason: reason);
        });
      }

      for (final (index, example) in invalidDateExamples.indexed) {
        final ((year, month, day), reason) = example;
        test('throws error for invalid date #$index', () {
          expect(() => LocalDate.of(year, month, day), throwsArgumentError,
              reason: reason);
        });
      }
    });

    group('creation from DateTime', () {
      test('succeeds', () {
        var dateTime = DateTime(2023, 11, 10);
        var localDate = LocalDate.fromDateTime(dateTime);
        expect(localDate, equals(LocalDate.of(2023, 11, 10)));

        dateTime = DateTime.utc(2023, 11, 10);
        localDate = LocalDate.fromDateTime(dateTime);
        expect(localDate, equals(LocalDate.of(2023, 11, 10)));
      });

      test('fails if the date is outside of the supported range', () {
        var dateTime = DateTime(0 - 1, 1, 1);
        expect(() => LocalDate.fromDateTime(dateTime), throwsArgumentError);

        dateTime = DateTime.utc(9999 + 1, 1, 1);
        expect(() => LocalDate.fromDateTime(dateTime), throwsArgumentError);
      });
    });

    // Parsing
    //--------------------------------------------------------------------------

    group('parsing', () {
      test('succeeds', () {
        final parsedDate =
            LocalDate.parse((2000, 10, 20), parser: TestDateFormat());

        expect(parsedDate, equals(LocalDate.of(2000, 10, 20)));
      });

      test('throws an exception if parser throws', () {
        expect(
            () => LocalDate.parse('invalid date format',
                parser: AlwaysThrowingDateParser<String>()),
            throwsFormatException);
      });

      test('throws an exception if parser returns invalid date components', () {
        expect(() => LocalDate.parse((-1, 1, 1), parser: TestDateFormat()),
            throwsFormatException);
        expect(() => LocalDate.parse((1, 0, 1), parser: TestDateFormat()),
            throwsFormatException);
        expect(() => LocalDate.parse((1, 1, 0), parser: TestDateFormat()),
            throwsFormatException);
      });
    });

    // Leap day determining
    //--------------------------------------------------------------------------
    test('leap day determining', () {
      expect(LocalDate.of(2000, 2, 29).isLeapDay, isTrue);
      expect(LocalDate.of(2000, 2, 29).isNotLeapDay, isFalse);

      expect(LocalDate.of(2000, 2, 28).isLeapDay, isFalse);
      expect(LocalDate.of(2000, 2, 28).isNotLeapDay, isTrue);

      expect(LocalDate.of(2000, 1, 29).isLeapDay, isFalse);
      expect(LocalDate.of(2000, 1, 29).isNotLeapDay, isTrue);
    });

    // Comparison tests
    //--------------------------------------------------------------------------

    test('equality & hash code', () {
      final reference = LocalDate.of(2000, 10, 15);
      expect(reference == reference, isTrue);

      final sameAsReference = LocalDate.of(2000, 10, 15);
      expect(reference, equals(sameAsReference));
      expect(reference.hashCode, equals(sameAsReference.hashCode),
          reason: 'Equal dates must have the same hash code.');

      expect(reference, isNot(LocalDate.of(2000, 10, 15 + 1)));
      expect(reference, isNot(LocalDate.of(2000, 10, 15 - 1)));
      expect(reference, isNot(LocalDate.of(2000, 10 + 1, 15)));
      expect(reference, isNot(LocalDate.of(2000, 10 - 1, 15)));
      expect(reference, isNot(LocalDate.of(2000 + 1, 10, 15)));
      expect(reference, isNot(LocalDate.of(2000 - 1, 10, 15)));
    });

    test('comparison with another local date', () {
      final reference = LocalDate.of(2000, 10, 20);

      expect(reference.compareTo(LocalDate.of(2000, 10, 20)), isZero);

      expect(reference.compareTo(LocalDate.of(2001, 1, 1)), isNegative);
      expect(reference.compareTo(LocalDate.of(1999, 12, 31)), isPositive);

      expect(reference.compareTo(LocalDate.of(2000, 11, 1)), isNegative);
      expect(reference.compareTo(LocalDate.of(2000, 9, 30)), isPositive);

      expect(reference.compareTo(LocalDate.of(2000, 10, 21)), isNegative);
      expect(reference.compareTo(LocalDate.of(2000, 10, 19)), isPositive);
    });

    test('is after another one', () {
      final reference = LocalDate.of(2000, 10, 20);

      expect(reference.isAfter(LocalDate.of(2000, 10, 19)), isTrue);
      expect(reference.isAfter(LocalDate.of(2000, 10, 20)), isFalse);
      expect(reference.isAfter(LocalDate.of(2000, 10, 21)), isFalse);

      expect(reference.isAfter(LocalDate.of(2000, 9, 30)), isTrue);
      expect(reference.isAfter(LocalDate.of(2000, 11, 1)), isFalse);

      expect(reference.isAfter(LocalDate.of(1999, 12, 31)), isTrue);
      expect(reference.isAfter(LocalDate.of(2001, 1, 1)), isFalse);
    });

    test('is before another one', () {
      final reference = LocalDate.of(2000, 10, 20);

      expect(reference.isBefore(LocalDate.of(2000, 10, 21)), isTrue);
      expect(reference.isBefore(LocalDate.of(2000, 10, 20)), isFalse);
      expect(reference.isBefore(LocalDate.of(2000, 10, 19)), isFalse);

      expect(reference.isBefore(LocalDate.of(2000, 11, 1)), isTrue);
      expect(reference.isBefore(LocalDate.of(2000, 9, 30)), isFalse);

      expect(reference.isBefore(LocalDate.of(2001, 1, 1)), isTrue);
      expect(reference.isBefore(LocalDate.of(1999, 12, 31)), isFalse);
    });

    test('is the same day as another', () {
      final reference = LocalDate.of(2000, 10, 20);

      expect(reference.isOnSameDayAs(LocalDate.of(2000, 10, 20)), isTrue);

      expect(reference.isOnSameDayAs(LocalDate.of(2000, 10, 19)), isFalse);
      expect(reference.isOnSameDayAs(LocalDate.of(2000, 10, 21)), isFalse);

      expect(reference.isOnSameDayAs(LocalDate.of(2000, 9, 20)), isFalse);
      expect(reference.isOnSameDayAs(LocalDate.of(2000, 11, 20)), isFalse);

      expect(reference.isOnSameDayAs(LocalDate.of(1999, 10, 20)), isFalse);
      expect(reference.isOnSameDayAs(LocalDate.of(2001, 10, 20)), isFalse);
    });

    test('is in the same year and month as another', () {
      final reference = LocalDate.of(2000, 10, 20);

      expect(reference.isInSameMonthAs(LocalDate.of(2000, 10, 20)), isTrue);
      expect(reference.isInSameMonthAs(LocalDate.of(2000, 10, 1)), isTrue);

      expect(reference.isInSameMonthAs(LocalDate.of(2000, 11, 20)), isFalse,
          reason: 'Different month.');
      expect(reference.isInSameMonthAs(LocalDate.of(2000, 9, 20)), isFalse,
          reason: 'Different month.');
      expect(reference.isInSameMonthAs(LocalDate.of(1999, 10, 20)), isFalse,
          reason: 'Different year.');
      expect(reference.isInSameMonthAs(LocalDate.of(2001, 10, 20)), isFalse,
          reason: 'Different year.');
    });

    test('is in the same year as another', () {
      final reference = LocalDate.of(2000, 10, 20);

      expect(reference.isInSameYearAs(LocalDate.of(2000, 10, 20)), isTrue);
      expect(reference.isInSameYearAs(LocalDate.of(2000, 1, 1)), isTrue);

      expect(reference.isInSameYearAs(LocalDate.of(1999, 10, 20)), isFalse);
      expect(reference.isInSameYearAs(LocalDate.of(2001, 10, 20)), isFalse);
    });

    // Formatting
    //--------------------------------------------------------------------------

    test('formatting', () {
      final date = LocalDate.of(2000, 10, 20);

      final formattedDate = date.formattedBy(TestDateFormat());
      expect(formattedDate, equals((2000, 10, 20)));
    });

    // Conversion to a different type
    //--------------------------------------------------------------------------

    test('conversion to string', () {
      var date = LocalDate.of(1, 2, 3);
      expect(date.toString(), equals('0001-02-03'));

      date = LocalDate.of(2023, 12, 31);
      expect(date.toString(), equals('2023-12-31'));
    });

    test('conversion to DateTime', () {
      final localDate = LocalDate.of(2023, 11, 10);

      var dateTime = localDate.toDateTime(utc: false);
      expect(dateTime, equals(DateTime(2023, 11, 10)));
      expect(dateTime.isUtc, isFalse);

      dateTime = localDate.toDateTime(utc: true);
      expect(dateTime, equals(DateTime.utc(2023, 11, 10)));
      expect(dateTime.isUtc, isTrue);
    });
  });
}


lib/src/year.dart
// Copyright (c) 2023 LLC "LitGroup"
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is furnished
// to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

import 'package:meta/meta.dart';

/// A representation of the year in the Gregorian calendar in the range from 0 to 9999.
@immutable
final class Year implements Comparable<Year> {
  // Construction
  //----------------------------------------------------------------------------

  /// Creates self from `int`.
  ///
  /// Throws [ArgumentError] if the [value] is out of supported range.
  static Year from(int value) {
    final year = tryFrom(value);
    if (year == null) {
      throw ArgumentError.value(
          value, 'value', 'Year is supported in range 0..9999.');
    }

    return year;
  }

  /// Creates self from `int`; returns null if the [value] is out of supported range.
  static Year? tryFrom(int value) {
    if (value < 0 || value > 9999) {
      return null;
    }

    return Year._internal(value);
  }

  Year._internal(this._value);

  final int _value;

  // Leap year determining
  //----------------------------------------------------------------------------

  /// Returns true if the year is leap.
  bool get isLeap =>
      (_value % 4 == 0) && (_value % 100 != 0 || _value % 400 == 0);

  /// Returns true if the year is not leap.
  bool get isNotLeap => !isLeap;

  /// Returns number of days in the year.
  ///
  /// 365 for a regular year and 366 for a leap one.
  int get durationInDays => isLeap ? 366 : 365;

  // Comparison
  //----------------------------------------------------------------------------

  bool isAfter(Year other) => _value > other._value;

  bool isBefore(Year other) => _value < other._value;

  bool isSameAs(Year other) => _value == other._value;

  @override
  int get hashCode => _value.hashCode;

  @override
  bool operator ==(Object other) => other is Year && isSameAs(other);

  @override
  int compareTo(Year other) => _value.compareTo(other._value);

  // Conversion to a different type
  //----------------------------------------------------------------------------

  /// Returns string representation of the year.
  ///
  /// Resulting string always containing 4 digits, used left padding by '0'.
  ///
  /// ```dart
  /// assert(Year.from(23).toString() == '0023');
  /// assert(Year.from(2023).toString() == '2023');
  /// ```
  @override
  String toString() => _value.toString().padLeft(4, '0');

  /// Returns integer representation of the year.
  int toInt() => _value;
}


lib/src/month.dart
// Copyright (c) 2023 LLC "LitGroup"
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is furnished
// to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

import 'package:meta/meta.dart';

import 'year.dart';

@immutable
final class Month implements Comparable<Month> {
  // Enumeration of values
  //----------------------------------------------------------------------------

  static const Month january = Month._internal(1, 'January');
  static const Month february = Month._internal(2, 'February');
  static const Month march = Month._internal(3, 'March');
  static const Month april = Month._internal(4, 'April');
  static const Month may = Month._internal(5, 'May');
  static const Month june = Month._internal(6, 'June');
  static const Month july = Month._internal(7, 'July');
  static const Month august = Month._internal(8, 'August');
  static const Month september = Month._internal(9, 'September');
  static const Month october = Month._internal(10, 'October');
  static const Month november = Month._internal(11, 'November');
  static const Month december = Month._internal(12, 'December');

  /// The list of month values.
  static const List<Month> values = [
    Month.january,
    Month.february,
    Month.march,
    Month.april,
    Month.may,
    Month.june,
    Month.july,
    Month.august,
    Month.september,
    Month.october,
    Month.november,
    Month.december,
  ];

  // Construction
  // ---------------------------------------------------------------------------

  /// Creates month from the number (from 1 to 12).
  ///
  /// Throws [ArgumentError] if the [number] if out of range.
  static Month from(int number) {
    return tryFrom(number) ??
        (throw ArgumentError('Invalid number of month ($number).'));
  }

  /// Creates month from the number (from 1 to 12);
  /// returns null if the number is invalid.
  static Month? tryFrom(int number) {
    if (number < 1 || number > 12) {
      return null;
    }

    return Month.values[number - 1];
  }

  const Month._internal(this._number, this._name);

  // Attributes
  //----------------------------------------------------------------------------

  /// The number of the month (from 1 to 12).
  final int _number;

  /// The name of the month (e.g. 'August').
  final String _name;

  /// Returns the number of days in the month.
  ///
  /// The [leapYear] parameter determines whether to return the length of the
  /// month in the leap or non-leap period.
  int days({required bool leapYear}) => switch (this) {
        Month.february => leapYear ? 29 : 28,
        Month.april || Month.june || Month.september || Month.november => 30,
        _ => 31
      };

  /// Returns the number of the month in the given year.
  int daysInYear(Year year) => days(leapYear: year.isLeap);

  // Comparison
  //----------------------------------------------------------------------------

  /// Returns `true` if this month is after the [other] in calendar order.
  bool isAfter(Month other) => _number > other._number;

  /// Returns `true` if this month is before the [other] in calendar order.
  bool isBefore(Month other) => _number < other._number;

  /// Returns `true` if this month is the same as [other].
  bool isSameAs(Month other) => _number == other._number;

  @override
  int get hashCode => _number;

  @override
  bool operator ==(Object other) => other is Month && isSameAs(other);

  @override
  int compareTo(Month other) => _number.compareTo(other._number);

  // Conversion to a different type
  //----------------------------------------------------------------------------

  /// Returns the name of the month (e.g. "January").
  @override
  String toString() => _name;

  /// Returns the calendar number of the month.
  int toInt() => _number;
}
